Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/TicketService.hs

```diff
@@ -594,3 +667,217 @@ postTicketBookingsUpdateSeats _ TicketBookingUpdateSeatsReq {..} = do
+  case ticketBookingServiceCategory.serviceCategoryId of
+    Nothing -> pure ()
+    Just id -> updateCancelledSeatsInSeatManagement (Kernel.Types.Id.Id id) totalNoOfCancelledSeats visitDate
+
+  QTBSC.updateSeatsAndAmountToRefundAndEventCancelledById (Just $ totalNoOfCancelledSeats + fromMaybe 0 ticketBookingServiceCategory.cancelledSeats) (ticketBookingServiceCategory.bookedSeats - totalNoOfCancelledSeats) (Just $ totalRefundAmount + fromMaybe 0 ticketBookingServiceCategory.amountToRefund) isEventCancelled ticketBookingServiceCategory.id
+  pure totalRefundAmount
+
+cancelTBSPeopleCategory :: Data.Time.Calendar.Day -> Kernel.Prelude.TimeOfDay -> API.Types.UI.TicketService.TicketBookingServicePeopleCategoryCancelReq -> Environment.Flow (HighPrecMoney, Int)
+cancelTBSPeopleCategory visitDate startTime tbsPeopleCategoryCancelReq = do
+  tBSPeopleCategory <- QTBPC.findById tbsPeopleCategoryCancelReq.id >>= fromMaybeM (PeopleCategoryNotFound tbsPeopleCategoryCancelReq.id.getId)
+  when (tbsPeopleCategoryCancelReq.quantity > tBSPeopleCategory.numberOfUnits || tbsPeopleCategoryCancelReq.quantity < 0) $ throwError (InvalidRequest "Invalid Quantity")
+
+  let noOfTicketToCancel = tbsPeopleCategoryCancelReq.quantity
+      noOfUnitAvailable = tBSPeopleCategory.numberOfUnits - noOfTicketToCancel
+
+  mbCancellationCharges <- case tBSPeopleCategory.peopleCategoryId of
+    Just peopleCategoryId -> do
+      peopleCategory <- QPC.findById peopleCategoryId
+      pure $ (.cancellationCharges) =<< peopleCategory
+    Nothing -> pure Nothing
+
+  cancellationCharges <- case mbCancellationCharges of
+    Just cancellationCharges -> calculateCancellationChargesForPeopleCategory tBSPeopleCategory noOfTicketToCancel cancellationCharges
+    Nothing -> pure $ HighPrecMoney 0
+  let amountToRefund = tBSPeopleCategory.pricePerUnit.amount * fromIntegral noOfTicketToCancel - cancellationCharges
+
+  QTBPC.updateTicketPeopleUnitsAndAmountById noOfUnitAvailable (Just noOfTicketToCancel) (Just $ amountToRefund + fromMaybe 0 tBSPeopleCategory.amountToRefund) tBSPeopleCategory.id
+  pure (amountToRefund, noOfTicketToCancel)
+  where
+    calculateCancellationChargesForPeopleCategory :: DTB.TicketBookingPeopleCategory -> Int -> [Domain.Types.ServicePeopleCategory.CancellationCharge] -> Environment.Flow HighPrecMoney
+    calculateCancellationChargesForPeopleCategory tBSPeopleCategory noOfTicketToCancel cancellationCharges' = do
+      currentTime <- getCurrentTime
+      let cancellationCharges = sortBy (\a1 a2 -> compare a2.time a1.time) cancellationCharges'
+          visitDateTime = UTCTime visitDate (timeOfDayToTime startTime)
+          remainingSeconds = Kernel.Prelude.roundToIntegral $ diffUTCTime currentTime visitDateTime
+          index = fromMaybe (length cancellationCharges - 1) $ findIndex (\item -> item.time DO.< remainingSeconds) cancellationCharges
```

Comment:
case NE.filter (\cancellationCharges_ -> cancellationCharges_.time <= remainingSeconds) $ sortBy (comparing (.time)) cancellationCharges' of
    [] -> -- No cancellation charge
    x -> last x -- fetch cancellation charge from this now

Reply:
@0utkarsh why [] -> -- No cancellation charge it will have no cancellation charge should it will have first element as cancellation charge

Reply:
It is a case match, if there is any element after filtering, it will go in 2nd case.